cmd/service-proxy/proxy.go

```diff
@@ -190,46 +191,69 @@ func (pm *ProxyManager) RoundTrip(req *http.Request) (resp *http.Response, err e
 	req.SetXForwarded()
 
-	l := pm.logger.WithValues("host", req.In.Host, "url", req.In.URL.String(), "method", req.In.Method)
+	// Create a logger with relevant request details
+	l := pm.logger.WithValues(
+		"incomingHost", req.In.Host,
+		"incomingRequestURL", req.In.URL.String(),
+		"incomingMethod", req.In.Method,
+	)
 
-	// inject current logger into context before returning
-	defer func() {
-		req.Out = req.Out.WithContext(log.IntoContext(req.Out.Context(), l))
-	}()
-
-	// hostname is expected to have the format specified in ./pkg/common/url.go
+	// Extract cluster from the incoming request host
 	cluster, err := common.ExtractCluster(req.In.Host)
 	if err != nil {
+		l.Error(err, "Failed to extract cluster from host", "host", req.In.Host)
+		req.Out = req.Out.WithContext(log.IntoContext(req.Out.Context(), l))
 		return
 	}
 
+	// Retrieve the upstream service route for the cluster
 	route, ok := pm.GetClusterRoute(cluster, "https://"+req.In.Host)
 	if !ok {
-		l.Info("No route found for cluster and URL", "cluster", cluster, "url", req.In.URL.String())
+		l.Info("No route found for cluster and URL", "cluster", cluster, "incomingRequestURL", req.In.URL.String())
+		req.Out = req.Out.WithContext(log.IntoContext(req.Out.Context(), l))
 		return
 	}
-	backendURL := route.url
-	// set cluster in context
-	ctx := context.WithValue(req.Out.Context(), contextClusterKey{}, cluster)
+	upstreamServiceRouteURL := route.url
 
-	l.WithValues("cluster", cluster, "namespace", route.namespace, "name", route.serviceName)
+	// Ensure the outgoing request URL is properly updated
+	if !strings.HasPrefix(req.Out.URL.Path, upstreamServiceRouteURL.Path) {
+		// Append the original request path to the upstream service route URL path
+		req.Out.URL.Path = strings.TrimSuffix(upstreamServiceRouteURL.Path, "/") + req.Out.URL.Path
+	}
+
+	// Set the correct upstream service route URL details
+	req.Out.URL.Scheme = upstreamServiceRouteURL.Scheme
+	req.Out.URL.Host = upstreamServiceRouteURL.Host
+
+	// Inject the cluster into the outgoing request context
+	ctx := context.WithValue(req.Out.Context(), contextClusterKey{}, cluster)
+	ctx = log.IntoContext(ctx, l)
 
 	req.Out = req.Out.WithContext(ctx)
-	req.SetURL(backendURL)
+
+	// Log the successful rewrite for debugging purposes
+	l.Info("Request rewrite completed",
+		"cluster", cluster,
+		"namespace", route.namespace,
+		"serviceName", route.serviceName,
+		"upstreamServiceRouteURL", req.Out.URL.String(),
+	)
 }
 
 // modifyResponse strips the k8s API server proxy path prepended to the location header during redirects:
 // https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/proxy/transport.go#L113
 func (pm *ProxyManager) modifyResponse(resp *http.Response) error {
+	logger := log.FromContext(resp.Request.Context())
+	logger.Info("Modifying response", "statusCode", resp.StatusCode, "originalLocation", resp.Header.Get("Location"))
+
```

Comment:
I am not very deep into how the whole service proxy works.  The comment on this method indicates that it was originally intended to handle the API server proxy path during redirects. 
cc @uwe-mayer

Reply:
>The comment on this method indicates that it was originally intended to handle the API server proxy path during redirects.

Yes, that is correct. I have tested and I can confirm that`modifyResponse` does handle redirects properly without any issues. I have also added tests for this method :)